aten/src/ATen/MemoryOverlap.cpp

```diff
@@ -3,24 +3,83 @@
 
 namespace at {
 
+struct SizeAndStride {
+  int64_t size;
+  int64_t stride;
+};
+
+/*
+ A comparator that will sort SizeAndStride structs by stride,
+ in ascending order.
+ */
+ int compareSizeAndStride(const void* a, const void* b) {
+  const SizeAndStride* aS = (const SizeAndStride*) a;
+  const SizeAndStride* bS = (const SizeAndStride*) b;
+
+  if (aS->stride < bS->stride) return -1;
+  if (aS->stride == bS->stride) return 0;
+  return 1;
+}
+
 MemOverlap has_internal_overlap(const Tensor& tensor) {
   return has_internal_overlap(tensor.unsafeGetTensorImpl());
 }
 
+/*
+Returns false if there is no possibility that the tensor
+has "overlapping" indices and true otherwise.
+"Overlapping" indices are two+ valid indices that specify
+the same offset within the tensor.
+The function does this by checking for a sufficient but not
+necessary condition of no overlap. In particular, that
+that there exists an ordering of the tensor's dimensions
+that is nicely "nested," with each dimension contained
+within the next one.
+*/
 MemOverlap has_internal_overlap(TensorImpl* t) {
   AT_ASSERT(t->layout() == kStrided);
 
   if (t->is_contiguous()) {
     return MemOverlap::NO;
   }
 
-  auto strides = t->strides();
-  if (strides.end() != std::find_if(
-        strides.begin(), strides.end(), [](int64_t s) { return s == 0; })) {
-    return MemOverlap::YES;
+  std::vector<SizeAndStride> info(t->dim());
+
+  int dims = t->dim();
```

Comment:
dims <= 1 tensor can't have overlaps, right?

Reply:
`x = torch.tensor(3.).expand(100)` has dim 1 and internal memory overlap

Reply:
(facepalm)